lib/waterline/query/dql/tenant.js

```diff
@@ -0,0 +1,193 @@
+
+  if (arguments.length === 0) {
+    //
+    // Error
+    return usageError('Tenant must be passed.', usage, cb);
+  }
+
+  if (typeof tenant === 'function') {
+    // Tenant is required
+    // Error
+    return usageError('Tenant must be passed.', usage, cb);
+  }
+
+
+  // Change connection settings based on tenant
+  // console.log('collections', this.waterline.collections);
+  // console.log('adapterDictionary', this.waterline, this.adapterDictionary);
+
+  var baseCollectionIdentity = self.baseIdentity || self.identity;
+  var connectionName = this.adapterDictionary['identity'];
+  var baseConnectionName = connectionName;
+  if (baseCollectionIdentity != self.identity) {
+      // Is not the base Collection
+      // Get base collection
+      var baseCollection = self.waterline.collections[baseCollectionIdentity];
+    //   console.log('baseCollection', baseCollectionIdentity);
+      baseConnectionName = baseCollection.adapterDictionary['identity'];
+    //   console.log('baseConnectionName', baseConnectionName);
+  }
+  var connection = this.connections[connectionName];
+  // Change for Multi-Tenancy support
+  // console.log('connection', connection);
+  var config = connection.config;
+  // console.log('config', config);
+
+  if (config.isMultiTenant && typeof config.configForTenant === "function") {
+
+    var configForTenant = config.configForTenant;
+
+    // Return Deferred or pass to adapter
+    if (typeof cb === 'function') {
+
+      var configDup = _.cloneDeep(config);
+      configDup.tenant = tenant; // Set Tenant
+      // Check if it is the BaseConnection
+      if (connectionName === baseConnectionName) {
+          // Reset the tenant on the connection for the global/main Collection
+          config.tenant = null;
+      }
+      // Get the config for this tenant
+      configForTenant.call(self, tenant, configDup, function(err, config) {
+        // console.log('configForTenant', err, config);
+        if (err) return cb(err);
+
+        // Tenant's config namespace
+        var tenantSuffix = "<tenant:" + tenant + ">";
+        var newConnectionIdentity = (baseConnectionName + tenantSuffix).toLowerCase();
+
+        function createTenantCollection() {
+
+            // console.log('create tenant collection', self instanceof require('../../collection'), self);
+            var newCollectionIdentity = (baseCollectionIdentity + tenantSuffix).toLowerCase();
+
+            // Get TenantCollection, if already exists
+            var tenantCollection = self.waterline.collections[newCollectionIdentity];
+            if (tenantCollection) {
+                // There is already an existing TenantCollection
+                return cb(null, tenantCollection);
+            }
+            // No existing TenantCollection found
+            // Create a new TenantCollection!
+
+            var WaterlineCollection = require('../../collection');
+            var colDupModel = extend.call(WaterlineCollection, {
```

Comment:
@particlebanana I think this is the only part that I am questionable on.

Objective: Create an extended copy of the base Collection, that uses the `newConnectionIdentity` as the `connection`.

Possible changes: Right now I believe I am ignoring custom class methods. Use [_.merge](https://lodash.com/docs#merge) to merge `{connection: [newConnectionIdentity]}` on top of the `base Collection` into a new object, thus creating a proper clone. I found that [_.cloneDeep](https://lodash.com/docs#cloneDeep) did not give me the desired result.

I will try to look at this today / tomorrow in preparation for your reviewing this weekend. I think this is the last thing I need to clean up. Let me know if you have any other ideas. Thanks.